packages/core/src/browser/messaging/ws-connection-provider.ts

```diff
@@ -61,28 +91,93 @@ export class WebSocketConnectionProvider extends AbstractConnectionProvider<WebS
             this.fireSocketDidClose();
         };
         socket.onmessage = ({ data }) => {
+            this.httpFallbackCounter = 0;
             this.handleIncomingRawMessage(data);
         };
         this.socket = socket;
         window.addEventListener('offline', () => this.tryReconnect());
         window.addEventListener('online', () => this.tryReconnect());
     }
 
+    handleSocketError(event: unknown): void {
+        this.httpFallbackCounter += 1;
+        if (this.httpFallbackOptions.allowed && this.httpFallbackCounter >= this.httpFallbackOptions.trigger) {
+            this.useHttpFallback = true;
+            this.socket.close();
+            const httpUrl = this.createHttpWebSocketUrl(WebSocketChannel.wsPath);
+            this.onHttpFallbackDidActivateEmitter.fire(undefined);
+            this.longPolling(httpUrl);
+            this.messageService().warn(
+                'Could not establish a websocket connection. The application will be using the HTTP fallback mode. This may affect performance and the behavior of some features.'
+            );
+        }
+        console.error(event);
+    }
+
+    async longPolling(url: string): Promise<void> {
+        let timeoutDuration = this.httpFallbackOptions.requestTimeout;
+        const controller = new AbortController();
+        const pollingId = window.setTimeout(() => controller.abort(), this.httpFallbackOptions.pollingTimeout);
+        try {
+            const response = await fetch(url, {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json'
+                },
+                signal: controller.signal,
+                keepalive: true,
+                body: JSON.stringify({ id: this.httpFallbackId, polling: true })
+            });
+            if (response.status === 200) {
+                window.clearTimeout(pollingId);
+                if (this.httpFallbackDisconnected) {
+                    this.fireSocketDidOpen();
+                }
+                const json: string[] = await response.json();
+                if (Array.isArray(json)) {
+                    for (const item of json) {
+                        this.handleIncomingRawMessage(item);
+                    }
+                } else {
+                    throw new Error('Received invalid long polling response.');
+                }
+            } else if (response.status >= 400) {
+                timeoutDuration = this.httpFallbackOptions.errorTimeout;
+                this.httpFallbackDisconnected = true;
+                this.fireSocketDidClose();
+                throw new Error('Response has no valid status code: ' + response.status);
```

Comment:
```suggestion
                throw new Error('Response has error code: ' + response.status);
```

'no valid status code' / 'invalid status code' makes it sound like it's not adhering to HTTP status codes, rather than that it sent a code that signifies an error.